### PR TITLE
lint: convert git-subtree-check.sh to Python

### PR DESCRIPTION
**Pull request by jacobpfickes**
Converts test/lint/git-subtree-check.sh to a Python script test/lint/git-subtree-check.py as requested in #24783

### CODE REVIEW

**Comment by fanquake** [comment]
https://github.com/bitcoin/bitcoin/pull/25039/checks?check_run_id=6237619420:
```bash
./ci/lint/06_script.sh: line 18: test/lint/git-subtree-check.sh: No such file or directory
```

**Comment by jacobpfickes** [comment]
@fanquake Sorry about that oversight. Fixed the issue and squashed. All checks pass now.

File: test/lint/git-subtree-check.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+    return parser.parse_args()
+
+def find_latest_squash(dir, commit):
+    sq = main = sub = None
+    stdout = subprocess.check_output(["git", "log", "--grep=^git-subtree-dir: " + dir + "/*$", "--pretty=format: START %H%n%s%n%n%b%nEND%n", commit], universal_newlines=True, encoding="utf8")
+    for line in stdout.splitlines():
+        if "START" in line:
+            sq = line.lstrip().split(' ')[1]
+        if "git-subtree-mainline:" in line:
```
Comment (laanwj) [inline test/lint/git-subtree-check.py:35]: `line.startswith("git-subtree-mainline:")`? (also the one below)

File: test/lint/git-subtree-check.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+            main = line.split(' ')[1]
```
Comment (laanwj) [inline test/lint/git-subtree-check.py:36]: `line.split(':', 2)[1].lstrip(' ')` maybe? (represents the intent of splitting header/value better, even if something is fudged with spaces)

File: test/lint/git-subtree-check.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    latest_squash = find_latest_squash(args.dir.rstrip('/'), args.commit)
+    if not latest_squash:
+        print(f"Error: {args.dir} is not a subtree")
+        sys.exit(2)
+    else:
+        old = latest_squash.split(' ')[0]
+        rev = latest_squash.split(' ')[1]
+
+    # get the tree in the current commit
+    tree_actual = subprocess.check_output(["git", "ls-tree", "-d", args.commit, args.dir, "| head -n 1"], universal_newlines=True, encoding="utf8")
```
Comment (laanwj) [inline test/lint/git-subtree-check.py:59]: whoops: `| head -n 1` is not going to work here, please implement this in Python

Comment (jacobpfickes) [reply test/lint/git-subtree-check.py]: fixed by piping output to new subprocess

File: test/lint/git-subtree-check.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    args = parse_args()
+
+    # find the latest subtree update
+    latest_squash = find_latest_squash(args.dir.rstrip('/'), args.commit)
+    if not latest_squash:
+        print(f"Error: {args.dir} is not a subtree")
+        sys.exit(2)
+    else:
+        old = latest_squash.split(' ')[0]
+        rev = latest_squash.split(' ')[1]
```
Comment (danielabrozzoni) [inline test/lint/git-subtree-check.py:56]: tiny nit: you can avoid putting `else` if you're exiting in the `if` branch:
```
    if not latest_squash:
        print(f"Error: {args.dir} is not a subtree")
        sys.exit(2)
    old = latest_squash.split(' ')[0]
    rev = latest_squash.split(' ')[1]
```
This reduces indentation, but here it's basically irrelevant :)

**Review by danielabrozzoni — APPROVED** [review]
light tACK 2d1d27afaefb050196e92e584170b20641668c9b - The code looks good, I tested locally and they produce the same output. 

(My tACK is light as I'm not really a bash expert, it's possible I missed some small differences between the two scripts)

File: test/lint/git-subtree-check.py (hunk 4 above)
Comment (jacobpfickes) [reply test/lint/git-subtree-check.py:56]: Very good point. I thought it helped with readability a bit but can easily go through and remove the else statements after exiting

Comment (danielabrozzoni) [reply test/lint/git-subtree-check.py:56]: As you wish, if you prefer to leave it like that it's ok for me :)

File: test/lint/git-subtree-check.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    return parser.parse_args()
+
+def find_latest_squash(dir, commit):
+    sq = main = sub = None
+    stdout = subprocess.check_output(["git", "log", "--grep=^git-subtree-dir: " + dir + "/*$", "--pretty=format: START %H%n%s%n%n%b%nEND%n", commit], universal_newlines=True, encoding="utf8")
+    for line in stdout.splitlines():
+        if "START" in line:
+            sq = line.lstrip().split(' ')[1]
+        if line.startswith("git-subtree-mainline:"):
+            main = line.split(':', 2)[1].split(' ')
```
Comment (laanwj) [inline test/lint/git-subtree-check.py:36]: I don't think this is correct. What I suggested was `.strip(' ')` not `.split(' ')`
(see https://github.com/bitcoin/bitcoin/pull/25039#discussion_r862685892)
Currently you end up with an array in `sub` which is likely not what you want.

**Comment by laanwj** [comment]
Thanks for addressing my comments.

I think the difficulty is properly testing this script. There's quite some logic in there that isn't exercised by the CI, and would require some different git aerobics to check.

File: test/lint/git-subtree-check.py (hunk 5 above)
Comment (jacobpfickes) [reply test/lint/git-subtree-check.py]: Sorry about that. You're correct. I completely misread it and just pushed a fix

**Comment by jacobpfickes** [comment]
> I think the difficulty is properly testing this script. There's quite some logic in there that isn't exercised by the CI, and would require some different git aerobics to check.

Would writing some sort of test script to hit all of the use cases be useful? and if so where is the best place to include something like that?

**Review by its0x08 — APPROVED** [review]

**Comment by laanwj** [comment]
> Would writing some sort of test script to hit all of the use cases be useful? and if so where is the best place to include something like that?

I don't know. From my point of view, testing the cases just once is fine, just to make sure the conditionals in the script are hit and do what they are supposed to.
Could publish it in a gist or similar.

I don't think there's need to run the test for this script say, as part of the CI every time.

**Comment by jacobpfickes** [comment]
Ok makes sense. I'll start putting together some of the commands used to hit all of the conditionals.

**Comment by jacobpfickes** [comment]
@laanwj 

I am currently trying to write a test guide but wanted to get your thoughts on a few of the conditionals. I am by no means a git expert so I might not be understanding correctly. I think with the way the script is written a couple of these failure conditionals will never be hit.

<br>
<br>

The following conditional produces an error if the subtree directory is not found in the commit.
https://github.com/bitcoin/bitcoin/blob/bb17f7233100d6660231d99fa5c2e2fc17a7747c/test/lint/git-subtree-check.py#L64-L69
But before it ever gets to that point it will try to find the latest squash (seen below) which will be empty if the directory is not found in the commit and will fail at that point
https://github.com/bitcoin/bitcoin/blob/bb17f7233100d6660231d99fa5c2e2fc17a7747c/test/lint/git-subtree-check.py#L35-L59

<br>
<br>
<br>

Also with the use of the following lines the `-d` flag will only return trees
https://github.com/bitcoin/bitcoin/blob/bb17f7233100d6660231d99fa5c2e2fc17a7747c/test/lint/git-subtree-check.py#L65-L66
https://github.com/bitcoin/bitcoin/blob/bb17f7233100d6660231d99fa5c2e2fc17a7747c/test/lint/git-subtree-check.py#L71-L72
which makes this conditional redundant
https://github.com/bitcoin/bitcoin/blob/bb17f7233100d6660231d99fa5c2e2fc17a7747c/test/lint/git-subtree-check.py#L74-L76

**Comment by laanwj** [comment]
Thanks for looking into it.
To be honest, I'm not really up to speed with git subtree internals either. You could very well be right! Seems the script could be simplified quite a bit, then (but out of scope for this PR).
I guess some of it is hardening to protect against `git` itself doing unexpected things. It's ok not to have a test plan for those cases.

**Comment by jacobpfickes** [comment]
Sure thing! Ok that makes sense. Yeah maybe it is checking for some outliers. I just wanted to make sure I wasn't missing something or someone with some more knowledge might know better. I wrap up how to trigger the other conditionals then.

**Comment by fanquake** [comment]
@jacobpfickes are you still working on this?

**Comment by jacobpfickes** [comment]
@fanquake yes, I am still working on this. Sorry I haven't had much free time lately. I should be able to dive back in this weekend though. Most of the development work is done but was just going to write a test plan to ensure its right and reviewers can easily validate

**Comment by achow101** [comment]
Are you still working on this?

**Comment by achow101** [comment]
Closing this as it has not had any activity in a while. If you are interested in continuing work on this, please leave a comment so that it can be reopened.

**Pull request closed without merging**